### ISSUE TITLE
Support for HP-UX

### ISSUE THREAD

**Issue by michaelbaeuerle** [fltk/fltk#937]
After the decision that FLTK 1.4 will still support autotools, this is the first attempt to get it working on an old HP machine with git revision 8b094e8af3215bed1c3c3c662e4b93163dac6bbc:
[FLTK-1.4_HP-UX_error_log.txt](https://github.com/fltk/fltk/files/14634312/FLTK-1.4_HP-UX_error_log.txt)

This OS does not support POSIX.1-2001 and provides no setenv().
The following patches use putenv(), if setenv() is not  available (code is from FLTK 1.3):
[patch_configh.in.txt](https://github.com/fltk/fltk/files/14634564/patch_configh.in.txt)
[patch_configure.ac.txt](https://github.com/fltk/fltk/files/14634568/patch_configure.ac.txt)
[patch_src_drivers_X11_Fl_X11_Screen_Driver.cxx.txt](https://github.com/fltk/fltk/files/14634718/patch_src_drivers_X11_Fl_X11_Screen_Driver.cxx.txt)
Todo: CMake needs such a check too or can define HAVE_SETENV unconditionally.

This patch fixes a namespace clash for _isspace():
[patch_src_Fl_Input.cxx.txt](https://github.com/fltk/fltk/files/14634594/patch_src_Fl_Input.cxx.txt)

This patch adds a missing file (required to build fluid):
[patch_src_Makefile.txt](https://github.com/fltk/fltk/files/14634634/patch_src_Makefile.txt)

vsscanf() is from C99. This OS provides a variant with non-const first parameter. Patch:
[patch_src_drivers_Unix_Fl_Unix_System_Driver.cxx.txt](https://github.com/fltk/fltk/files/14634734/patch_src_drivers_Unix_Fl_Unix_System_Driver.cxx.txt)

roundf() is from C99. This patch tries to emulate it with floor():
[patch_nanosvg_nanosvg.h.txt](https://github.com/fltk/fltk/files/14634774/patch_nanosvg_nanosvg.h.txt)
Because it is only used for colors, it should be acceptable that the patch does not exactly reproduce the behaviour of roundf().

This patch is for a visibility problem in Fl_Grid:
[patch_FL_Fl_Grid.H.txt](https://github.com/fltk/fltk/files/14634455/patch_FL_Fl_Grid.H.txt)
It simply makes the affected members public. I have tried to fix the problem with "friend" declarations without success.
If nobody knows a better fix, the "public" declaration should be limited to old compilers.

With the patches above FLTK 1.4 works on this machine.
I have run all test programs and anything looked OK.

**Comment by Albrecht-S** [comment]
@michaelbaeuerle Thank you very much for this elaborate report. I like your report style, showing the error message and the fix for each one separately. Comments in the patches are an extra bonus. :+1:

I looked into your patches, and everything looks plausible, although I may change some of the patches. For instance, we try not to patch bundled libs (nanosvg) although we did that and have our own fork. I'll probably move the inclusion of nanosvg to a separate file (which I had considered before anyway) and add your new code (patch) before the inclusion of the nanosvg header, or something like that.

I'll probably add every patch in a separate commit.

**Comment by Albrecht-S** [comment]
Todo list:
- [ ] [patch_configh.in.txt](https://github.com/fltk/fltk/files/14634564/patch_configh.in.txt)
- [ ] [patch_configure.ac.txt](https://github.com/fltk/fltk/files/14634568/patch_configure.ac.txt)
- [ ] [patch_src_drivers_X11_Fl_X11_Screen_Driver.cxx.txt](https://github.com/fltk/fltk/files/14634718/patch_src_drivers_X11_Fl_X11_Screen_Driver.cxx.txt)
- [ ] CMake needs such a check too or can define HAVE_SETENV unconditionally.
- [ ] [patch_src_Fl_Input.cxx.txt](https://github.com/fltk/fltk/files/14634594/patch_src_Fl_Input.cxx.txt)
- [ ] [patch_src_Makefile.txt](https://github.com/fltk/fltk/files/14634634/patch_src_Makefile.txt)
- [ ] Consider adding this file to CMake as well
- [ ] [patch_src_drivers_Unix_Fl_Unix_System_Driver.cxx.txt](https://github.com/fltk/fltk/files/14634734/patch_src_drivers_Unix_Fl_Unix_System_Driver.cxx.txt)
- [ ] [patch_nanosvg_nanosvg.h.txt](https://github.com/fltk/fltk/files/14634774/patch_nanosvg_nanosvg.h.txt)
- [ ] [patch_FL_Fl_Grid.H.txt](https://github.com/fltk/fltk/files/14634455/patch_FL_Fl_Grid.H.txt)

**Comment by ManoloFLTK** [comment]
@albrecht: I've already committed at 5a97629 a fix regarding the unresolved call to XUtf8LookupString() and stop here regarding this issue.

**Comment by Albrecht-S** [comment]
@michaelbaeuerle Regarding [patch_src_drivers_Unix_Fl_Unix_System_Driver.cxx.txt](https://github.com/fltk/fltk/files/14634734/patch_src_drivers_Unix_Fl_Unix_System_Driver.cxx.txt):
```
+#if defined(__hpux)
+  // HP-UX 11.11 provides:
+  // int vsscanf(char *s, const char *format, va_list ap);
+  int retval = vsscanf((char*)input, format, args);
+#else  // defined(__hpux)
   int retval = vsscanf(input, format, args);
+#endif  // defined(__hpux)
```

I wonder what happened if a newer version of HP-UX provided the standard (`const`) version of `vsscanf().` Wikipedia lists newer versions from `11.20 (2001) – 11i v1.5` up to `11.31 (2007) – 11i v3`. Citation:
> This release [HP-UX 11i v3] supports both PA-RISC and Itanium. HP-UX 11i v3 conforms to [The Open Group](https://en.wikipedia.org/wiki/The_Open_Group)'s UNIX 03 [standard](https://en.wikipedia.org/wiki/Single_UNIX_Specification).

Some of these releases were (according to Wikipedia) target at IA-64 only which we can likely neglect, but the latest relase was for "both PA-RISC and Itanium".

Questions:
- What do you know about these newer releases?
- Should we modify the patch?
- I assume that the opposite "conversion" - from `'char*'` to `'const char*'` would be harmless and issue a warning rather than an error, hence we could use the patch as-is?

**Comment by Albrecht-S** [comment]
@ManoloFLTK Thanks for commit 5a97629d0b272765b607dca1757ef067328f9d55

**Comment by wcout** [comment]
Regarding `roundf()` in nanosvg:

nanosvg is a bit inconsistent, as it using `roundf()` in `nanosvg.h`, but has a own function defined for it in `nanosvgrast.h`:

`static float nsvg__roundf(float x) { return (x >= 0) ? floorf(x + 0.5) : ceilf(x - 0.5); }`

which seems to be more accurate than the definition in the patch.

**Comment by michaelbaeuerle** [comment]
Albrecht Schlosser wrote:
> 
> @michaelbaeuerle Regarding [patch_src_drivers_Unix_Fl_Unix_System_Driver.cxx.txt](https://github.com/fltk/fltk/files/14634734/patch_src_drivers_Unix_Fl_Unix_System_Driver.cxx.txt):
> ```
> +#if defined(__hpux)
> +  // HP-UX 11.11 provides:
> +  // int vsscanf(char *s, const char *format, va_list ap);
> +  int retval = vsscanf((char*)input, format, args);
> +#else  // defined(__hpux)
>    int retval = vsscanf(input, format, args);
> +#endif  // defined(__hpux)
> ```
> 
> I wonder what happened if a newer version of HP-UX provided the
> standard (`const`) version of `vsscanf().`
> Wikipedia lists newer versions [...]
> 
> Some of these releases were (according to Wikipedia) target at IA-64
> only which we can likely neglect, but the latest relase was for
> "both PA-RISC and Itanium".
> 
> Questions:
> - What do you know about these newer releases?

I have never used a newer version, but I would expect them to be POSIX
conformant (regardless of CPU architecture).

My version simply predates POSIX.1-2001 and provides vsscanf() as a
non-standard extension.

> - Should we modify the patch?
> - I assume that the opposite "conversion" - from `'char*'` to
>   `'const char*'` would be harmless and issue a warning rather than
>   an error, hence we could use the patch as-is?

I agree. It should work for newer HP-UX versions too.
But yes, it looks a bit like a hack (and indeed is a hack).

The clean solution would be to let the build systems check whether the
system provides the POSIX variant of vsscanf():
<https://pubs.opengroup.org/onlinepubs/9699919799/functions/vsscanf.html>
and always use it, if available.
A similar mechanism is already present for scandir() from POSIX.1-2008.

I can implement a solution for autoconf, if you want to solve it this
way. CMake can simply define HAVE_VSSCANF unconditionally (not suitable
for such old systems anyway).

**Comment by michaelbaeuerle** [comment]
wcout wrote:
> 
> Regarding `roundf()` in nanosvg:
> 
> nanosvg is a bit inconsistent, as it using `roundf()` in `nanosvg.h`,
> but has a own function defined for it in `nanosvgrast.h`:
> 
> `static float nsvg__roundf(float x)
>     { return (x >= 0) ? floorf(x + 0.5) : ceilf(x - 0.5); }`
> 
> which seems to be more accurate than the definition in the patch.

Thank you for pointing out, I have only looked in "nanosvg.h".

Looks like that even the FLTK fork of nanosvg cannot simply use it.
Because "nanosvgrast.h" includes "nanosvg.h", nsvg__roundf() should
be defined in "nanosvg.h" to be usable in both files.

**Comment by Albrecht-S** [comment]
@wcout Thanks for pointing this out. I was just going to implement the nanosvg patch (separation in its own file is already done). My first thought was that we might just '#define roundf nsvg__roundf' but I'll check how we can implement this. ... 

Although I found a working solution I'll likely patch nanosvg instead and make a PR upstream because it's a little weird to do it:
```
#include <math.h>                   // must be before #define below !
static float nsvg__roundf(float x); // prototype (see nanosvgrast.h)
#define roundf nsvg__roundf         // redefinition (#937)
```

Works, but "not nice" (TM).

**Comment by Albrecht-S** [comment]
Regarding vsscanf(): @michaelbaeuerle wrote:
> I agree. It should work for newer HP-UX versions too.
[...] The clean solution would be to let the build systems check whether the system provides the POSIX variant of vsscanf() ...

OK, I'll go for the simple solution (use your patch as-is). It's unlikely that someone uses a newer HP-UX version, and if, then there will be only a compiler warning. Only HP-UX users would be concerned anyway. Adding a build system check would be worthless for likely 99.999% of our users. ;-)

**Comment by Albrecht-S** [comment]
@michaelbaeuerle wrote:
> Looks like that even the FLTK fork of nanosvg cannot simply use it. Because "nanosvgrast.h" includes "nanosvg.h", nsvg__roundf() should be defined in "nanosvg.h" to be usable in both files.

My current version "works" by using the #define as written above. One problem with an upstream patch would be if someone wanted to separate the implementation in two files (you could do this!) because `nsvg__roundf()` is `static`. Hence it can be implemented only once if used in one compilation unit but would be needed in both compilation units otherwise. This looks like needing another "hack", maybe something like
```
#ifndef NSVG_ROUNDF_DEFINED
static float nsvg__roundf(float x) { return (x >= 0) ? floorf(x + 0.5) : ceilf(x - 0.5); }
#define NSVG_ROUNDF_DEFINED
#endif
```
... in both files or something similar. Better ideas?

Side note: an unnamed namespace might help as well (would it?) but we don't allow using this in FLTK yet.

FTR: I can commit my patch anyway and then think about patching our nanosvg fork - and submitting a PR - later.

**Comment by Albrecht-S** [comment]
@michaelbaeuerle Now I'm confused. You wrote:
> The problem with this replacement function is that it still is not sufficient for C90. C90 provides floor() and ceil(), this means the double-variants. The float-variants with "f"-suffix are from C99 too.

What compiler (or libc) dialect does your HP-UX system provide? The replacement function always used `floorf()` and `ceilf()` in `nanosvgrast.h` but you didn't provide a patch for this. Does your system "understand" C99?

My current patch doesn't change this usage in `nanosvgrast.h`. I'm puzzled.

That said, **my primary goal is to provide C++98 and C99 compatibility**. My test system builds FLTK with these compatibility settings w/o warnings. I believe that's all we can achieve w/o major changes for all such subtle differences like floor() vs. floorf() etc..

Being compatible with C90 would be a bonus but I wouldn't want to invest too much for systems that are very, very likely older than 25 years, even if **_one or two of them_** ;-) are still using FLTK. If they can use FLTK 1.3, then so be it, for all other systems we're providing FLTK 1.4, and in FLTK 1.5 there will be even more restrictions. Time and development is proceeding...

**Comment by Albrecht-S** [comment]
FTR: `nanosvg` update pushed in commit 1d43ae0b0a01fdd90486c616d5dddd318a8fb9f0. This should fix this partial issue for C99 and later. Hopefully it's enough for the target system (HP-UX 11.11).

I intend to file an upstream PR for nanosvg later.

**Comment by Albrecht-S** [comment]
I believe it's all done now. For a list of tasks and related commits please see https://github.com/fltk/fltk/issues/937#issuecomment-2004037560.

@michaelbaeuerle Please confirm that the latest version works on the target system (HP-UX 11.11).

Commit 5af2d77b8468bc585ac350b53d704f030b74fe1d (Fl_Grid) is different than your proposed patch but I hope it fixes the problems.

If there are any issues left, please feel free to add comments.

**Comment by michaelbaeuerle** [comment]
Albrecht Schlosser wrote:
> 
> @michaelbaeuerle Now I'm confused. You wrote:
> > The problem with this replacement function is that it still is not
> sufficient for C90. C90 provides floor() and ceil(), this means the
> double-variants. The float-variants with "f"-suffix are from C99 too.  
> 
> What compiler (or libc) dialect does your HP-UX system provide?
> The replacement function always used `floorf()` and `ceilf()` in
> `nanosvgrast.h` but you didn't provide a patch for this.
> Does your system "understand" C99?

The compiler was GCC 4.0.1 (and has support for C99). But the POSIX
API and libc from the OS are C90 based.

> My current patch doesn't change this usage in `nanosvgrast.h`.
> I'm puzzled.

Most compilers provide features beyond the standards that they fully
support.

> That said, **my primary goal is to provide C++98 and C99 compatibility**.
> My test system builds FLTK with these compatibility settings w/o
> warnings. I believe that's all we can achieve w/o major changes for all
> such subtle differences like floor() vs. floorf() etc..

As long as it works, there is no need to fix something.

> Being compatible with C90 would be a bonus but I wouldn't want to
> invest too much for systems that are very, very likely older than
> 25 years, even if **_one or two of them_** ;-) are still using FLTK.
> If they can use FLTK 1.3, then so be it, for all other systems we're
> providing FLTK 1.4, and in FLTK 1.5 there will be even more
> restrictions. Time and development is proceeding...

I appreciate that you added the patches required for HP-UX.

**Comment by Albrecht-S** [comment]
@michaelbaeuerle Thanks for your comments. I take it that your HP-UX system works with the latest commit.

Did you already have a chance to test it, and can we close this issue now?

**Comment by michaelbaeuerle** [comment]
Albrecht Schlosser wrote:
> 
> @michaelbaeuerle Thanks for your comments. I take it that your HP-UX
> system works with the latest commit.
> 
> Did you already have a chance to test it, and can we close this issue now?

Not yet. I hope I can test HP-UX again tonight or tomorrow.

**Comment by michaelbaeuerle** [comment]
Tested again and it works now.
Can be closed.
